middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -208,150 +323,407 @@ let get_exn_extra_args env k =
+      (* for duplicated bindings, we need to replace the original splittable
+         binding with the new split binding in the bindings map of the env *)
+      match split_binding.inline with
+      | Must_inline_once -> env
+      | Must_inline_and_duplicate ->
+        { env with
+          bindings = Variable.Map.add var (Binding split_binding) env.bindings
+        }
+    in
+    let env =
+      List.fold_left
+        (fun env new_binding ->
+          let flambda_var = Variable.create "to_cmm_tmp" in
+          { env with
+            bindings = Variable.Map.add flambda_var new_binding env.bindings
+          })
+        env new_bindings
+    in
+    will_inline_complex env res split_binding
+
+let pop_from_top_stage ?consider_inlining_effectful_expressions env var =
+  match env.stages with
+  | [] -> None
+  | Effect var_from_stage :: prev_stages ->
+    (* In this case [var_from_stage] corresponds to an effectful binding forming
+       the most recent stage. We also know that [var] doesn't have an available
+       pure defining expression (either because that expression isn't pure, or
+       because the corresponding binding has already been flushed). As such, we
+       can't move the defining expression for [var] past that of
+       [var_from_stage], in the case where these variables are different.
+       However if these two variables are in fact the same, we can consider
+       inlining the defining expression. *)
+    let consider_inlining_effectful_expressions =
+      match consider_inlining_effectful_expressions with
+      | Some consider -> consider
+      | None -> Flambda_features.Expert.inline_effects_in_cmm ()
+    in
+    if Variable.equal var var_from_stage
+       && consider_inlining_effectful_expressions
+    then Some { env with stages = prev_stages }
+    else None
+  | Coeffect_only vars_from_stage :: prev_stages ->
+    (* Here we see if [var] has a coeffect-only defining expression on the most
+       recent stage. If so, then we can commute it with any other expression on
+       the stage, since they all only have coeffects. The defining expression
+       for [var] may then be considered for inlining. *)
+    if Variable.Set.mem var vars_from_stage
+    then
+      let new_vars_in_stage = Variable.Set.remove var vars_from_stage in
+      let stages =
+        if Variable.Set.is_empty new_vars_in_stage
+        then prev_stages
+        else Coeffect_only new_vars_in_stage :: prev_stages
+      in
+      Some { env with stages }
+    else None
+
+let inline_variable ?consider_inlining_effectful_expressions env res var =
+  match Variable.Map.find var env.bindings with
+  | exception Not_found -> (
+    (* this happens for continuation parameters and bindings that have been
+       flushed *)
+    match Variable.Map.find var env.vars with
+    | exception Not_found ->
+      Misc.fatal_errorf "Variable %a not found in env" Variable.print var
+    | e -> e, env, res, Ece.pure_can_be_duplicated)
```

Comment:
This could do with a comment, explaining that `env.vars` only stores expressions of the form `Cvar v` so the effects are always `pure_can_be_duplicated`

Reply:
comment added